internal/state/graph/httproute.go

```diff
@@ -401,24 +402,93 @@ func findAcceptedHostnames(listenerHostname *v1beta1.Hostname, routeHostnames []
 		return []string{hostname}
 	}
 
-	match := func(h v1beta1.Hostname) bool {
-		if hostname == "" {
-			return true
-		}
-		return string(h) == hostname
-	}
-
 	var result []string
 
 	for _, h := range routeHostnames {
-		if match(h) {
-			result = append(result, string(h))
+		routeHost := string(h)
+		if match(hostname, routeHost) {
+			result = append(result, GetMoreSpecificHostname(hostname, routeHost))
 		}
 	}
 
 	return result
 }
 
+func match(listenerHost, routeHost string) bool {
+	if listenerHost == "" {
+		return true
+	}
+
+	if routeHost == listenerHost {
+		return true
+	}
+
+	wildcardMatch := func(host1, host2 string) bool {
+		return strings.HasPrefix(host1, "*.") && strings.HasSuffix(host2, strings.TrimPrefix(host1, "*"))
+	}
+
+	// check if listenerHost is a wildcard and routeHost matches
+	if wildcardMatch(listenerHost, routeHost) {
+		return true
+	}
+
+	// check if routeHost is a wildcard and listener matchess
+	return wildcardMatch(routeHost, listenerHost)
+}
+
+// GetMoreSpecificHostname returns the more specific hostname between the two inputs.
+//
+// This function assumes that the two hostnames match each other, either:
+// - Exactly
+// - One as a substring of the other
+// - Both as substrings of some parent wildcard
+func GetMoreSpecificHostname(hostname1, hostname2 string) string {
+	if hostname1 == hostname2 {
+		return hostname1
+	}
+
+	if hostname1 == "" {
+		return hostname2
+	} else if hostname2 == "" {
```

Comment:
nit: don't need the else here
```suggestion
	} 
	if hostname2 == "" {
```

Reply:
Not sure I see much reason to switch to a blank if statement, except for preference.

Reply:
IMHO it's harder to read. If everything is left aligned it is easier to scan the code more quickly. 

https://medium.com/@matryer/line-of-sight-in-code-186dd7cdea88

Reply:
I'm surprised this was missed by the linter: https://github.com/mgechev/revive/blob/master/RULES_DESCRIPTIONS.md#superfluous-else. This linter must be limited to `else` blocks